Initial set of API Standards

dgmstuart opened the pull request:
These are largely based on the White House and Google standards, as well as on our conversation on Thursday 12th October

Please check that everything is basically sane and as we discussed. If significant additions need to be made then these can be done as further PRs

dgmstuart committed:
Create initial set of draft API standards

These were originally written into a Google Doc, and now ported into
markdown.

It would have been nice to just refer to a single external document, but
the whitehouse standards miss out some fundamental specifics, and the
google standards go into too much specific detail.

dgmstuart committed:
Specify the Create a Repair endpoint in detail

dgmstuart committed:
Specify endpoint for retrieving a repair

Should be identical to creating a repair

dgmstuart commented:
Would be good to get a review from Andre as well but he hasn't been added to this github yet

dgmstuart commented on README.md:
````diff
@@ -189,3 +189,92 @@ GET /v1/properties/:propertyReference/
+Create a repair request, with or without a list of work orders
+
+```
+POST /v1/properties/:id/repairs
````
Note: This URL is an idea @leeky and I have been discussing. 

It makes sense on the basis that internal repairs are always for a property, and one can imagine `GET /v1/properties/:property_reference/repairs` returning all the repairs for a particular property at some point in the future. 

However, individual repairs don't need to be fetched in this scoped way: we only need to pass a repair reference:

```
GET /v1/repairs/:repair_request_reference
```
 and not the property reference as well - e.g.
```
GET /v1/properties/:property_reference/repairs/:repair_request_reference
```

I don't think we need/want to add that restriction.

LBHAKovalichin replied:
This makes sense. If you already have a reference of a specific object you should not need to specify another.

LBHakairamkonda reviewed (commented):
Shouldn't this be similar to getting the property details by property reference?
```json
 {
  "results": [{
    "requestReference": "08912445",
   "propertyReference": "00078345",
   "problemDescription": "The fan is buzzing and sometimes not spinning at all",
  "priority": "N",
  "contact": {
    "name": "Al Smith",
    "telephoneNumber": "07876543210",
    "emailAddress": "al.smith@hotmail.com",
    "callbackTime": "8am - 12pm"
  },
   "workOrders": [
     {
       "workOrderReference": "20090190",
       "sorCode":  "20090190",
       "supplierReference": "00000127"
     }
   ]
}
   ]
 }

dgmstuart commented:
@LBHakairamkonda Which place did you mean? You can comment directly on the code by clicking the blue [+] icon when hovering over a line of code.

dgmstuart committed:
Rationale for returning resource in POST responses

Discussed this with Andrey and Leeky. The Microsoft convention is to
return just an ID, the convention of some well-regarded open source APIs
is to return the whole object.

There's no explicit standard around this, and REST doesn't describe what
should happen, beyond saying that the created resource must be
identifiable from the response.

leeky commented on README.md:
```diff
@@ -2,3 +2,338 @@
+  - Where just a date is required, still provide a full time object, with the time as 00:00:00
+  - Example time: `"2017-11-06T16:34:41Z"`
+  - Example date: `"2017-11-06T00:00:00Z"`
+- Durations should be represented as a number of minutes:
```
[ISO8601](https://en.wikipedia.org/wiki/ISO_8601#Durations) also defines a standard for durations, not sure if we want to consider using it, or stick with minutes for simplicity's sake.

dgmstuart replied:
We discussed this and concluded that it looks like a standard which is comprehensive, but overengineered for our use case.

LBHAKovalichin replied:
For hackney purposes any duration can be represented in minutes. The standard includes years and months and across services I do not see any duration that will exceed a few days. (Even if it does current convention is to use From and To dates for those)

leeky commented on README.md:
```diff
@@ -2,3 +2,338 @@
+## Data Properties
+
+- Property names should be in lower camel case
+  - Good: `"contactTelephoneNumber"`
```
Should these be left unquoted?

e.g. `contactTelephoneNumber` vs `"contactTelephoneNumber"`

dgmstuart replied:
Yup - probably better: I've added a new commit

leeky commented on README.md:
```diff
@@ -2,3 +2,338 @@
+- Dates and Times should be formatted as RFC3339:
+  - Don’t include milliseconds
+  - Always use UTC (represented with Z)
+  - Where just a date is required, still provide a full time object, with the time as 00:00:00
```
:+1:

dgmstuart replied:
Yeah probably better.

dgmstuart replied:
Done

leeky reviewed (commented):
Just a couple of small comments, otherwise looks pretty good.

dgmstuart committed:
Remove quotes from attribute name examples

dgmstuart commented:
@LBHakairamkonda If you could clarify your comment, I can address and get this merged.

LBHakairamkonda reviewed (commented):
Looks good to me

LBHakairamkonda reviewed (commented):
Looks good to me
